clients/hadoopfs/src/main/java/io/lakefs/storage/PhysicalAddressTranslator.java

```diff
@@ -13,16 +15,18 @@ public PhysicalAddressTranslator(String blockstoreType, String validityRegex) {
         this.validityRegex = validityRegex;
     }
 
-    // translate a URI in lakeFS storage namespace syntax into a Hadoop FileSystem URI
-    public URI translate(URI uri) throws URISyntaxException {
-        if(!Pattern.compile(validityRegex).matcher(uri.toString()).find()) {
-            throw new RuntimeException(String.format("URI %s does not match blockstore namespace regex %s", uri.toString(),
+    // translate a physical address in lakeFS storage namespace syntax into a Hadoop FileSystem Path
+    public Path translate(String address) throws URISyntaxException {
+        if(!Pattern.compile(validityRegex).matcher(address).find()) {
+            throw new RuntimeException(String.format("address %s does not match blockstore namespace regex %s", address,
             validityRegex));
         }
+        
+        // Going through Path.toUri to avoid encoding bugs. See: https://github.com/treeverse/lakeFS/issues/5827
+        URI uri = new Path(address).toUri();
```

Comment:
It's cool that this works!  But: let's open an issue to refactor this method.

1. It does a workaround to create the URI
1. It performs validity checking _without_ looking at the URI.
1. It only supports S3.
1. All it ends up doing is replacing "s3://" with "s3a://", and verifying that the URL is valid.  But the URL _should_ be valid, and we'd anyway get a failure at the very next step if it were invalid.

I mean, in Perl you could just do this with
```perl
$address =~ s/^s3:/s3a:/ or die 'only S3 URLs supported'
```

Of course this will be contentious, and no point in holding up this very good fix to go through all that nonsense.  So... can we open an issue please?

Reply:
Can you please explain what is missing in this fix? What do we want to change in the new issue?